x/btcstaking/keeper/power_dist_change.go

```diff
@@ -56,31 +56,61 @@ func (k Keeper) UpdatePowerDist(ctx context.Context) {
 	// to construct the new distribution
 	newDc := k.ProcessAllPowerDistUpdateEvents(ctx, dc, events, maxActiveFps)
 
-	// find newly bonded finality providers and execute the hooks
-	newBondedFinalityProviders := newDc.FindNewActiveFinalityProviders(dc)
-	for _, fp := range newBondedFinalityProviders {
-		if err := k.hooks.AfterFinalityProviderActivated(ctx, fp.BtcPk); err != nil {
-			panic(fmt.Errorf("failed to execute after finality provider %s bonded", fp.BtcPk.MarshalHex()))
-		}
-	}
-
 	// record voting power and cache for this height
-	k.recordVotingPowerAndCache(ctx, newDc)
+	k.recordVotingPowerAndCache(ctx, dc, newDc, maxActiveFps)
 	// record metrics
 	k.recordMetrics(newDc)
 }
 
-func (k Keeper) recordVotingPowerAndCache(ctx context.Context, dc *types.VotingPowerDistCache) {
+func (k Keeper) recordVotingPowerAndCache(ctx context.Context, prevDc, newDc *types.VotingPowerDistCache, maxActiveFps uint32) {
+	if prevDc == nil {
+		panic("the previous voting power distribution cache cannot be nil")
+	}
+
+	// deep copy the previous dist cache if the new dist cache is nil
+	if newDc == nil {
+		newDc = types.NewVotingPowerDistCache()
+		newDc.TotalVotingPower = prevDc.TotalVotingPower
+		newDc.NumActiveFps = prevDc.NumActiveFps
+		newFps := make([]*types.FinalityProviderDistInfo, len(prevDc.FinalityProviders))
+		for i, prevFp := range prevDc.FinalityProviders {
+			newFp := *prevFp
+			newFps[i] = &newFp
+		}
+		newDc.FinalityProviders = newFps
+	}
```

Comment:
Why do we need a deep copy here? It seems that `prevDc` is only used in L104. Is
it possible to change L103-L110 to that, if `newDc != nil` then execute
L103-L110, otherwise avoid doing so? To make L85-L91 work we can simply do
`newDc := prevDc`

Reply:
if we don't do deep copy, the change at Line 90 will change the `IsTimestamped` value in `prevDc`